Numerical Domains (Interval, Octagons)

gitsimon opened the pull request:
The interval/octagonal domain work. 

Details:

- Numerical base classes restructured
- Devision of Octagons/Intervals into basic value, lattice and domain consistently structured and named
- Octagon closure fixed to also infer unary constraints over two and more 'hops'
- Closure does not cancels widening
- Fixed warnings where possible

gitsimon committed:
setup numerical domains

gitsimon committed:
setup sphinx docs

gitsimon committed:
inter

gitsimon committed:
refactor lattice: separate TOP/BOTTOM special elements with mixins

gitsimon committed:
refactor lattice: remove clumsy and verbose element constructor, use default always

also change examples to directly set element (mutable lattices!) not reassigning new objects

gitsimon committed:
fix dbm structure, add dict-like generators

gitsimon committed:
inter (problems)

gitsimon committed:
inter

gitsimon committed:
try another closure algorithm (test pass!)

Algorithm from paper: An Improved Tight Closure Algorithm for Integer Octagonal Constraints - Roberto
        Bagnara, Patricia M. Hill, Enea Zaffanella

gitsimon committed:
add expression visitor and transformer (adoption from ast module)

gitsimon committed:
inter

gitsimon committed:
inter

gitsimon committed:
restructure generic test cases, mutable IntervalLattice

tests pass

gitsimon committed:
refactorings

gitsimon committed:
octagonal assignments

and homogenize the index tuple order for octagon

gitsimon committed:
factor out octagonal assignment cases

gitsimon committed:
fix some bugs

gitsimon committed:
fix graphviz problem with empty <FONT>...</FONT> blocks

solution: add space at the end

gitsimon committed:
octagon assignments works

gitsimon committed:
more bug fixes

gitsimon committed:
comments from meeting

gitsimon committed:
inter

gitsimon committed:
semantics connection

gitsimon committed:
sample

gitsimon committed:
fix enum fields not allowed to be defined inside class

gitsimon committed:
refactor linear forms

gitsimon committed:
complete octagons (without fallbacks)

gitsimon committed:
expression simplifier with unittests

gitsimon committed:
come closer to working octagons

gitsimon committed:
fix octagon set to BOTTOM if underlying CDBM is inconsistent

gitsimon committed:
correct missunderstanding of ExpressionTransformer

always call visit (not generic_visit) to ensure proper handling of current node!

gitsimon committed:
bug fix

gitsimon committed:
rename samples

gitsimon committed:
fix bug in unittest result check

gitsimon committed:
fix bug

gitsimon committed:
close octagons for unittest and display

gitsimon committed:
more samples and tests

gitsimon committed:
assume fallback to interval

but interval assume is not yet implemented!

gitsimon committed:
assignment fallback to interval & bug fix (sign of lb)

gitsimon committed:
ignore non-int vars

gitsimon committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	abstract_domains/generic_lattices.py
#	abstract_domains/lattice.py
#	abstract_domains/liveness/livedead.py
#	abstract_domains/usage/store.py
#	requirements.txt
#	tests/playground.py
#	visualization/graph_renderer.py

gitsimon committed:
remove default methods

set default in __init__ if not already done

caterinaurban commented on abstract_domains/dummies.py:
```diff
@@ -0,0 +1,71 @@
+    def __init__(self):
+        self._expr = None
+        super().__init__()
+        self.bottom()
```
Here the BOTTOM element is misused (and possibly the TOP element as well).

caterinaurban commented on abstract_domains/numerical/test_IntervalStore.py:
```diff
@@ -0,0 +1,36 @@
+
+
+if __name__ == '__main__':
+    suite()
```
I would move this into the `tests` or `unittests` module.

caterinaurban commented on abstract_domains/numerical/test_Simplifier.py:
```diff
@@ -0,0 +1,35 @@
+
+
+if __name__ == '__main__':
+    suite()
```
I would move this into the `tests` or `unittests` module.

caterinaurban commented on abstract_domains/numerical/test_SingleVarLinearForm.py:
```diff
@@ -0,0 +1,42 @@
+
+
+if __name__ == '__main__':
+    suite()
```
I would move this into the `tests` or `unittests` module.

caterinaurban commented on abstract_domains/numerical/test_dbm.py:
```diff
@@ -0,0 +1,133 @@
+
+
+if __name__ == '__main__':
+    suite()
```
I would move this into the `tests` or `unittests` module.

caterinaurban commented on abstract_domains/numerical/numerical.py:
```diff
@@ -0,0 +1,61 @@
+    @classmethod
+    def evaluate(cls, expr: Expression):
+        """Evaluates the expression within this abstract domain."""
+        raise NotImplementedError()
```
Why not make it an abstract method as well?

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,242 @@
+from core.expressions_tools import ExpressionVisitor
+
+
+class IntervalLattice(BoundedLattice, NumericalLattice):
```
I think that this class is mixing two things that should not be mixed. IMHO a single interval as an abstraction of a set of values is separate from an interval as a lattice element. The class IntervalLattice should have an Interval as an element (where Interval implements add, sub, etc.).

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,242 @@
+                raise ValueError(f"Variable type {expr.typ} is not supported!")
+
+
+class IntervalDomain(IntervalStore, State):
```
I do not understand why IntervalStore and IntervalDomain are separate. I would expect that IntervalDomain is a Store lattice that can be used as a State.

caterinaurban commented on abstract_domains/numerical/numerical.py:
```diff
@@ -0,0 +1,61 @@
+
+
+# rename (without lattice) e.g. IntegerOperations
+class NumericalLattice(metaclass=ABCMeta):
```
Isn't this class representing sets of values and the operations on them? I would maybe rename this class NumericalValue or just Value. The class Interval would inherit from it. Alternatively, we just drop this class entirely and just keep the class Interval.

caterinaurban reviewed (commented):
I did not have time to review the whole code. I added just a few comments to discuss tomorrow.

gitsimon committed:
add space to avoid error when label empty

gitsimon committed:
fix bug

Loop Header was a Basic instead Loop node

gitsimon committed:
fix dangerous shared initialization

and make consistent optional param

gitsimon committed:
extend numeric interface

gitsimon committed:
while sample and make it work!

gitsimon committed:
move tests to unittests folder

gitsimon committed:
separate TOP/BOTTOM lattice mixins

gitsimon committed:
refactorings

interval domain inheritance structure
octagon uses only BottomMixin
renamings

gitsimon committed:
add explicit BOTTOM to used and used_liststart domain

this allows correction of store.is_bottom() test from all to any of the store entries equals BOTTOM is enough to let store be BOTTOM

gitsimon committed:
double fix! DBM closure must not operate on a copy

this bug prevented transitive constraints to be propagated!
samples are corrected, some optimization (not traverse upper right triangular part of DBM)

gitsimon committed:
remove stale TODOs

gitsimon committed:
State inherits from Lattice now (not BoundedLattice)

gitsimon committed:
move interval/octagon conversion methods

gitsimon committed:
cleanup

fix warnings and related stuff

gitsimon committed:
Merge remote-tracking branch 'upstream/master'

gitsimon committed:
expression expander

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -164,35 +164,67 @@ def replace(self, other: 'Lattice') -> 'Lattice':
 
-class BoundedLattice(Lattice, metaclass=ABCMeta):
-    """A mutable lattice element, with default bottom and top elements."""
+class Kind(Enum):
```
IMHO this belongs inside KindMixin (as it was inside BoundedLattice before)

gitsimon replied:
done

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -164,35 +164,67 @@ def replace(self, other: 'Lattice') -> 'Lattice':
-        DEFAULT = 2
-        BOTTOM = 1
+class KindMixin(Lattice, metaclass=ABCMeta):
+    """A Mixin to add a capability to distinguish between user-defined and special elements like TOP/BOTTOM."""
```
I would rather write """A mixin that adds an explicit distinction between default and special lattice elements like bottom/top."""

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -164,35 +164,67 @@ def replace(self, other: 'Lattice') -> 'Lattice':
 
+
+class TopMixin(KindMixin, metaclass=ABCMeta):
+    """A mixin that adds a TOP element to another lattice."""
```
I would write `top` instead of `TOP` (do not capitalize)

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -164,35 +164,67 @@ def replace(self, other: 'Lattice') -> 'Lattice':
         self._kind = kind
 
+
+class TopMixin(KindMixin, metaclass=ABCMeta):
```
No need for ABCMeta here

gitsimon replied:
It is needed whenever a class does intentionally not define all abstract methods

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -164,35 +164,67 @@ def replace(self, other: 'Lattice') -> 'Lattice':
-        TOP = 3
-        DEFAULT = 2
-        BOTTOM = 1
+class KindMixin(Lattice, metaclass=ABCMeta):
```
No need for ABCMeta here

gitsimon replied:
It is needed whenever a class does intentionally not define all abstract methods

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -164,35 +164,67 @@ def replace(self, other: 'Lattice') -> 'Lattice':
+
+
+class BottomMixin(KindMixin, metaclass=ABCMeta):
+    """A mixin that adds a BOTTOM element to another lattice."""
```
I would write bottom instead of BOTTOM (do not capitalize)

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -164,35 +164,67 @@ def replace(self, other: 'Lattice') -> 'Lattice':
+        return self._kind == Kind.TOP
+
+
+class BottomMixin(KindMixin, metaclass=ABCMeta):
```
No need for ABCMeta here

gitsimon replied:
It is needed whenever a class does intentionally not define all abstract methods

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -164,35 +164,67 @@ def replace(self, other: 'Lattice') -> 'Lattice':
+        return self._kind == Kind.BOTTOM
+
+
+class BoundedLattice(TopMixin, BottomMixin, metaclass=ABCMeta):
```
No need for ABCMeta here

Also, nitpicking, I personally prefer to have BottomMixin before TopMixin. There is no real difference code-wise, but it makes more sense to me from a semantic point of view.

gitsimon replied:
It is needed whenever a class does intentionally not define all abstract methods

caterinaurban commented on abstract_domains/lattice.py:
```diff
@@ -164,35 +164,67 @@ def replace(self, other: 'Lattice') -> 'Lattice':
+
+    def __init__(self):
+        super().__init__()
+        self._kind = Kind.DEFAULT
```
I think you can remove this `__init__` since it is inherited from `KindMixin`

gitsimon replied:
true

caterinaurban commented on unittests/interval/assignments.py:
```diff
@@ -0,0 +1,9 @@
+x = 3
+y = 5
+
+a = x+y
```
When you run the interval analysis here `x -> [8,8]` instead of `x -> [3,3]`

gitsimon replied:
cannot reproduce (fixed in the meanwhile?)

caterinaurban replied:
It is fixed now ;)

caterinaurban commented on unittests/interval/assignments.py:
```diff
@@ -0,0 +1,9 @@
+
+if a > 0:
+    a = 2*a
+
```
Here is would expect `a -> [16,16]` instead of `a -> [8, 16]`. The test `not(a > 0)` should result in the bottom interval for `a`.

gitsimon replied:
assume is not yet implemented for interval domain, TODO is added

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+            l = self.visit(expr.left)
+            r = self.visit(expr.right)
+            if expr.operator == BinaryArithmeticOperation.Operator.Add:
+                return l.add(r)
```
Careful here! When you evaluate you should not modify the current interval (see comments on assignments.py, for instance).

gitsimon replied:
make a copy when a element is retrieved from store

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+        self._upper = None
+        super().__init__()
+        self._lower = lower
+        self._upper = upper
```
Why not just

```
assert lower <= upper
super().__init__()
self._lower = lower
self._upper = upper
```

?

gitsimon replied:
done

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+        return self
+
+    def negate(self):
+        self.interval = (min(-self.lower, -self.upper), max(-self.lower, -self.upper))
```
I do not understand what you are doing here. I would expect `self.interval = (-self.upper, -self.lower)` here.

gitsimon replied:
TBD

gitsimon replied:
fixed

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+
+    @upper.setter
+    def upper(self, b):
+        self._upper = b
```
To be consistent with all the other class definitions, I would move these last four definitions right after `___init___` and before `interval`

gitsimon replied:
done

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+        self._upper = b
+
+    def __repr__(self):
+        return f"[{self.lower},{self.upper}]"
```
To be consistent with all the other class definitions, I would move this right before `add`

gitsimon replied:
done

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+class IntervalLattice(Interval, BottomMixin):
+    def __repr__(self):
+        if self.is_bottom():
+            return "BOTTOM"
```
Can we use the unicode character `U+22A5` instead?

gitsimon replied:
done

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+        return f"[{self.lower},{self.upper}]"
+
+
+class IntervalLattice(Interval, BottomMixin):
```
Don't you need to override `is_bottom` to check that `upper < lower`? Can you accidentally turn bottom intervals into non-bottom intervals, e.g., `[1, 0] + [1, 3] = [2, 3]`?

gitsimon replied:
This is a problem, yes. Should you even be able to calculate with a bottom interval? then the subclassing of interval is wrong and it should be aggregation

gitsimon replied:
we should add emptyness property to intervals and synchronize this with bottom property of interval lattice

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+    class Visitor(ExpressionVisitor):
+        """A visitor to abstractly evaluate an expression (without variables) in the interval domain."""
+
+        # noinspection PyMethodOverriding
```
Why don't you use a matching signature `def generic_visit(self, expr, *args, **kwargs)` instead?

gitsimon replied:
because the code gets very unreadable, since then I have to 'parse' the arguments inside the method, a functionality which is much better handled by explicit parameter names

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+                f"{type(self)} does not support generic visit of expressions! "
+                f"Define handling for expression {type(expr)} explicitly!")
+
+        # noinspection PyMethodMayBeStatic,PyUnusedLocal
```
Why don't you change the signature to `def visit_Input(self, _: Input)` instead of suppressing the PyUnusedLocal warning?

gitsimon replied:
done

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+            if left.typ == int:
+                self.store[left] = self._visitor.visit(right)
+        else:
+            raise NotImplementedError("")
```
Can you add a more meaningful error message?

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+        return self
+
+    def _assume(self, condition: Expression) -> 'IntervalDomain':
+        return self
```
This does not seem right (see comments on assignments.py, for instance). Is it just temporary? At least add a `TODO`

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+        return self  # nothing to be done
+
+    def _substitute_variable(self, left: Expression, right: Expression):
+        raise NotImplementedError("")
```
Please add a meaningful error message.

caterinaurban commented on abstract_domains/numerical/numerical.py:
```diff
@@ -0,0 +1,27 @@
+class NumericalMixin(metaclass=ABCMeta):
+    @abstractmethod
+    def forget(self, var: VariableIdentifier):
+        """Forget all constraints about a variable."""
```
constraints -> information

caterinaurban commented on abstract_domains/numerical/numerical.py:
```diff
@@ -0,0 +1,27 @@
+
+    @abstractmethod
+    def set_bounds(self, var: VariableIdentifier, lower: int, upper: int):
+        """Sets the upper and lower bound of a variable.
```
Sets -> Set

caterinaurban commented on abstract_domains/numerical/numerical.py:
```diff
@@ -0,0 +1,27 @@
+
+    @abstractmethod
+    def get_bounds(self, var: VariableIdentifier):
+        """Returns the upper and lower bound of a variable.
```
Returns -> Return

caterinaurban commented on abstract_domains/numerical/numerical.py:
```diff
@@ -0,0 +1,27 @@
+
+    @abstractmethod
+    def evaluate(self, expr: Expression):
+        """Evaluates the expression within this abstract domain."""
```
Evaluates -> Evaluate
abstract -> numerical

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+    _visitor = Visitor()  # static class member shared between all instances
+
+
+class IntervalDomain(Store, NumericalMixin, State):
```
I would rename the file `interval_domain`

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+        return self.to_interval_domain().evaluate(expr)
+
+
+class OctagonDomain(OctagonLattice, State):
```
I would rename the file `octagon_domain`

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+        super().__init__()
+
+        for key in self.dbm.keys():
+            self.dbm[key] = inf
```
Do you need this? Doesn't the initialization of a dbm already do this?

gitsimon replied:
true

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+            self._index_to_var[index + 1] = var
+            index += 2
+        self._dbm = IntegerCDBM(len(variables) * 2)
+        super().__init__()
```
Why not this first?

gitsimon replied:
done

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+        self._var_to_index = {}
+        self._index_to_var = {}
+        index = 0
+        for var in self.variables:
```
It would be nice to have some (line) comments to explain what is going on here. For instance, with these two lines:
```
self._index_to_var[index] = var
self._index_to_var[index + 1] = var
```

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+        if self.is_bottom():
+            return "BOTTOM"
+        elif self.is_top():
+            return "TOP"
```
Why don't we use the unicode characters `U+22A5` (bottom) and `U+22A4` (top) instead?

gitsimon replied:
done

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+
+Sign = UnaryArithmeticOperation.Operator
+PLUS = Sign.Add
+MINUS = Sign.Sub
```
Why are you reusing the operators of a unary arithmetic operation in this context? It also seems a bit hacky to have this defined at the top level like this.

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+                            res.append(f"-{var1.name}+{var2.name}≤{c}")
+                        c = self[PLUS, var1, MINUS, var2]
+                        if c < inf:
+                            res.append(f"-{var1.name}-{var2.name}≤{c}")
```
I do not really understand how this business with `PLUS` and `MINUS` works. Maybe comments would help.

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+
+    def _less_equal(self, other: 'OctagonLattice') -> bool:
+        if self.dbm.size != other.dbm.size:
+            raise ValueError("Can not compare octagons with unequal sizes!")
```
Can not -> Cannot

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+    return 1 if sign == MINUS else 0
+
+
+class OctagonLattice(BottomMixin, NumericalMixin):
```
This class desperately needs docstrings. I would also rename the methods that are not supposed to be used from the outside with an underscore in front.

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+        return self
+
+    def forget(self, var: VariableIdentifier):
+        # close first to not loose implicit constraints about other variables
```
loose -> lose

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+                    BinaryComparisonOperation(expr.typ, expr.left, BinaryComparisonOperation.Operator.GtE, expr.right))
+                return ConditionSet([cs1.conditions[0], cs2.conditions[0]], ConditionSet.Operator.MEET)
+
+            elif expr.operator == BinaryComparisonOperation.Operator.Eq:
```
This condition is the same of line 339

gitsimon replied:
is a bug, now fixed

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+
+
+class OctagonDomain(OctagonLattice, State):
+    class SmallerEqualConditionTransformer(ExpressionTransformer):
```
This class also desperately needs docstrings

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+        super().__init__(variables)
+
+    def _substitute_variable(self, left: Expression, right: Expression) -> 'OctagonDomain':
+        raise NotImplementedError()
```
Meaningful error?

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+
+        # noinspection PyMethodOverriding
+        def visit(self, expr, state):
+            return super().visit(expr, state)
```
This is also a bit ugly since we have to pass the state. Do we really need to have this as a subclass?

caterinaurban commented on abstract_domains/numerical/octagon.py:
```diff
@@ -0,0 +1,578 @@
+        self._assign_negated_other_var(x, y)
+        self._assign_same_var_plus_constant(x, interval)
+
+    def _assign_variable(self, left: Expression, right: Expression) -> 'OctagonDomain':
```
Meaningful error messages?

caterinaurban reviewed (commented):
I think the octagon.py file needs *a lot* more docstrings, comments, meaningful error messages. If you do not want to do this now, at least add TODOs.

gitsimon committed:
variadic expression support in linear forms

gitsimon committed:
make expression expander generic

gitsimon committed:
refactoring

rename expressions to operands

gitsimon committed:
improve formatting of operation expressions

do not add unnecessary brackets
adapt tests

gitsimon committed:
use expression expander to rewrite simplifier

and slightly adapt tests
refactor expression visitor/transformer finalize_result

gitsimon committed:
refactorings

move Kind

caterinaurban commented on abstract_domains/numerical/interval.py:
```diff
@@ -0,0 +1,237 @@
+
+        def visit_VariableIdentifier(self, expr: VariableIdentifier):
+            if expr.typ == int:
+                return self.store.store[expr]
```
`deepcopy` here?

gitsimon committed:
comments from 17.7

gitsimon committed:
intervals can be empty

add checks and synchronize with lattice bottom property

gitsimon committed:
renamings, typos

gitsimon committed:
fix octagon NotEq

add test

gitsimon committed:
typos, cleanup

gitsimon committed:
octagon docs

gitsimon committed:
Merge remote-tracking branch 'upstream/master' into upstream-pull-request

gitsimon commented:
The review comments are incorporated or TODO comments added to code.

gitsimon committed:
separate linear form visitor to inner class

gitsimon committed:
fix bugs, add rich comparison operators to interval, linear form

caterinaurban commented on abstract_domains/numerical/interval_domain.py:
```diff
@@ -0,0 +1,307 @@
+            raise NotImplementedError("Incomparable types!")
+        if self.empty() or other.empty():
+            raise NotImplementedError("Empty intervals are not comparable!")
+        return self.lower >= other.upper
```
Do we really need all these comparison operators? They are also confusing since they have a different semantics from what I would expect since I think of intervals as sets of integers. Are we sure that this is the semantic that we want?

caterinaurban commented on core/expressions.py:
```diff
@@ -547,6 +572,19 @@ def __str__(self):
             elif self.value == 10:
                 return "not in"
 
+    Operator.REVERSE_OPERATOR = {
```
Why is this here and not inside the subclass Operator? This causes an unresolved attribute reference warning in reverse_operator.

gitsimon replied:
see other comment

caterinaurban commented on core/expressions.py:
```diff
@@ -525,6 +543,13 @@ class Operator(BinaryOperation.Operator):
+        def reverse_operator(self):
+            """Returns the reverse operator of this operator."""
+            try:
+                return BinaryComparisonOperation.Operator.REVERSE_OPERATOR[self]
```
Unresolved attribute reference 'REVERSE_OPERATOR' for class 'Operator'

gitsimon replied:
I don't see any warnings for unresolved attribute and the code is correct (covered by tests).
the line
`Operator.REVERSE_OPERATOR=...`
adds that attribute to the Operator inner class; this construction is necessary because otherwise the `REVERSE_OPERATOR` field would get interpreted as an Enum constant

caterinaurban commented on unittests/test_Expander.py:
```diff
@@ -0,0 +1,38 @@
+import unittest
+
+from core.expressions_tools import simplify, expand
```
Unused import statement: simplify

gitsimon replied:
removed

caterinaurban commented on core/special_expressions.py:
```diff
@@ -0,0 +1,41 @@
+
+    def __str__(self):
+        string_list = [f"({str(operand)})" if isinstance(operand, Operation) else str(operand) for operand in
+                       self.operands]
```
Expected 'collections.Iterable', got 'Union[List[Expression], None, list]' instead

gitsimon replied:
Weakness of type inference:
`self._operands = operands or []` seems not to be handled correctly (always results in a iterable if `operands` is an iterable or None

I fixed the type declaration in the constructor however.

caterinaurban commented on docs/conf.py:
```diff
@@ -0,0 +1,159 @@
+# General information about the project.
+project = 'Lyra'
+copyright = '2017, Caterina Urban, Simon Wehrli, Mostafa Abdullah'
+author = 'Caterina Urban, Simon Wehrli, Mostafa Abdullah'
```
Mostafa Hassan

gitsimon replied:
corrected

caterinaurban commented on abstract_domains/numerical/interval_domain.py:
```diff
@@ -0,0 +1,307 @@
+
+    def _less_equal(self, other: 'IntervalLattice') -> bool:
+        # NOTE: do not use less equal operator of plain interval since that has different semantics (every value in
+        # interval is less equal than any value in other interval)
```
I would remove the less equal operator that you added to the Interval class and remove this comment as well.

caterinaurban commented on abstract_domains/numerical/linear_forms.py:
```diff
@@ -0,0 +1,186 @@
+        if not isinstance(other, self.__class__):
+            raise NotImplementedError("Incomparable types!")
+        return self.var_summands == other.var_summands \
+               and self.interval >= other.interval
```
I guess these are the ones using the comparisons in Interval. But the we need these? And again, are we sure about the semantics?

gitsimon replied:
It is a very valid question, if one form _is definitely smaller than another form_. If we actually use it, as discussed in last meeting, maybe not

caterinaurban reviewed (commented):
just one concern about comparison of intervals and a couple of warnings

gitsimon committed:
Merge remote-tracking branch 'upstream/master' into upstream-pull-request

# Conflicts:
#	abstract_domains/traces/traces_domain.py

gitsimon committed:
small fixes
